Restricted compute tactics in Elpi and Ltac2

MSoegtropIMC opened the pull request:
This adds restricted compute functions (full cbv under application of given head symbol) written in Elpi and Ltac2.

A few notes:

- the Elpi variant requires 1.6.0 (which is newer than Coq Platform) - you might want to keep it anyway for comparison
- the Ltac2 variant has an extra type check (call to `check_throw`) which can be commented out for performance checks
- the head symbol for the Ltac2 variant is given as `constr` - this means that if you specify functions with implicit arguments you need to give an @ - otherwise the term cannot possibly be the head of an application since it has implicit arguments applied to it
- the Ltac2 varaint has a "limited recusion" and a "full recursion" variant, which defer in what kind of structures they recurse into looking for the application - see the documentation. The "limited reduction" variant is faster and usually what you want - the other is more complete. It might make sense to make specially tailored variants by removing cases from the full version.
- in my tests the performance of the elpi and ltac2 version (without extra type check) are about the same.

MSoegtropIMC commented:
Thanks for testing and the report!

What is going wrong has nothing to do with the reduction (cbv), but with taking the complete term apart and putting it together again. What the recursion for searching applications does is chopping the term into individual pieces, recurse and possibly modify the pieces and put the pieces together again. If the term would be Lego bricks, the "full recursion" variant destructs it into each individual piece and puts it together again. The "limited reduction" variant keeps larger substructures intact. Obviously there is something going wrong in the process - I need to investigate what exactly causes the issue.

Meanwhile I extended the "limited reduction" variant to look into match expressions, but not into match returns and match case functions (this can be added). This leads to the reduction you want in the example case. Looking into match expressions definitely makes sense - I am not so sure about the match case functions. For performance reasons the "limited reduction" variant should look into (and take apart + reconstruct) as little as possible while still getting the job done. If you feel you need to look into the case functions, please let me know - or just take the "full recursion" code as reference and experiment with extending the "limited recursion" variant.

It might btw. make sense to either test upfront or return information on if any of the subterms changed and if not return the original term. But first I am not sure this would be faster - the destruction / reconstruction should be very fast and second doing the reconstruction always is much more likely to find issues early (as in the example you shared where there is no change at all).

I am looking into the full reduction variant and see if I can find what is going wrong.

One more note: after experimenting with what works for us, the goal is to have this implemented in OCaml. The issue is that nobody really knows what kind of simplification / specialized computation makes sense for automation as in VST.

andrew-appel commented:
I have tested the cbv_uao_lr tactic, and indeed it is significantly faster than my previous way of doing this using `repeat (set (a := f _); compute in a; subst a)`.

I suggest the following minor changes, and then I will be happy to merge this pull request.  Normally I would just merge and then make the changes myself, but this particular P.R. is the development of a tactic that may also be useful in other applications such as VST, so I would prefer that the original author coordinate these improvements in a single place.

1.  The "Tests / examples" at the end should not need to import ZArith and Lia (which is rather heavyweight); it should be easy to reformulate them just using Nat.
2. Change the comment "fixpoints, matches and notably types are copied unchanged" to just "fixpoints and types are copied unchanged"
3. Put a warning at the beginning of `eval_cbv_uao_fr` that the tactic is known to have bugs.  This warning could be a comment in the file, or even better, a call to `idtac` with a warning message.
4. Consider, perhaps, removing the elpi version from this P.R. until we see what version of elpi is included with the next Coq Platform.
5. Add an Ltac-callable tactic as follows:
```
Ltac compute_every := 
  (* compute_every f   will find every term below the line 
    of the form  (f _) or (f _ _) etc. whose head is f, 
   and fully reduce it using "compute". 
  This tactic does NOT look for (f _) within types or fixpoints.
*)
  ltac2:(f |- compute_tactics_ltac2.cbv_uao_lr (Option.get (Ltac1.to_constr f))).
```

MSoegtropIMC commented:
I had a look at the anomaly. Here is a fairly minimal reproduction:
```
Require Import Ltac2.Ltac2.
Require Import Ltac2.Printf.

Ltac2 message_of_option (pr : 'a -> message) (ov : 'a option) :=
  match ov with
  | Some v => Message.concat (Message.of_string "Some ") (pr v) 
  | None => Message.of_string "None"
  end.

Ltac2 check_throw (term : constr) : constr :=
  match Constr.Unsafe.check term with
  | Val c => c
  | Err e => Control.throw e
  end.

Ltac2 redflags_full () :=
{
  Std.rBeta := true;
  Std.rDelta := true;
  Std.rMatch := true;
  Std.rFix := true;
  Std.rCofix := true;
  Std.rZeta := true;
  Std.rConst := []
}.

Ltac2 Notation "dbg_print" t(thunk(tactic(6))) := t ().

Ltac2 rec eval_cbv_uao_fr (head : constr) (rf : Std.red_flags) (term : constr) : constr :=
  match Constr.Unsafe.kind term with
  | Constr.Unsafe.Prod binder bound =>
      dbg_print printf "eval_cbv_uao_fr Prod =>";
      let binder_c := eval_cbv_uao_fr_binder head rf binder in
      let bound_c := eval_cbv_uao_fr head rf bound in
      let kind_c := Constr.Unsafe.Prod binder_c bound_c in
      let result := Constr.Unsafe.make kind_c in
      dbg_print printf "eval_cbv_uao_fr Prod <=";
      result

  | _ => term
  end

with eval_cbv_uao_fr_binder (head : constr) (rf : Std.red_flags) (b : binder) : binder := 
  dbg_print printf "eval_cbv_uao_fr_binder %a =>" (fun _ => message_of_option Message.of_ident) (Constr.Binder.name b);
  let name := Constr.Binder.name b in
  let type := Constr.Binder.type b in
  dbg_print printf "eval_cbv_uao_fr_binder A";
  let type_c := eval_cbv_uao_fr head rf type in
  dbg_print printf "eval_cbv_uao_fr_binder B";
  let result := Constr.Binder.make name type_c in
  dbg_print printf "eval_cbv_uao_fr_binder <=";
  result
  .

Ltac2 cbv_uao_fr (head : constr) : unit :=
    let goal := Control.goal() in
    let goal_c := eval_cbv_uao_fr head (redflags_full ()) goal in
    (* The line below can be commented out for performance tests - it is just an extra debug type check*)
    let goal_c := check_throw goal_c in
    change $goal_c.

Goal forall (A:Type) (a:A), a=a.
cbv_uao_fr 'Nat.mul.
```
The print trace is:
```
eval_cbv_uao_fr Prod =>
eval_cbv_uao_fr_binder Some A =>
eval_cbv_uao_fr_binder A
eval_cbv_uao_fr_binder B
eval_cbv_uao_fr_binder <=
eval_cbv_uao_fr Prod =>
eval_cbv_uao_fr_binder Some a =>
eval_cbv_uao_fr_binder A
eval_cbv_uao_fr_binder B
```
So it fails to reconstruct the dependent binder `(a : A)`. Even if I don't compute in the type (use `type` instead of `type_c` it fails. The issue here is that - according to some discussions on Zulip I had on this a while back - the function `Constr.Binder.make` does typecheck the term, which obviously cannot work since `A` is a dangling reference.

In master there is an additional function `unsafe_make` to get around this (the function `Constr.Unsafe.make` is unsafe in this sense as well). See (https://github.com/coq/coq/blob/c686e93e52bf7a6307bf450733cc7616a6244d7d/user-contrib/Ltac2/Constr.v#L109).

Until we have access to this function there is not that much I can do about this, except avoiding running into such cases by limited recursion. One can't easily detect this and one can't catch anomalies.

This all looks a bit bleeding edge, but then if we don't start to get things done with Elpi and Ltac2, issues won't get fixed.

MSoegtropIMC commented:
@andrew-appel : I now changed the tactics as follows:

- don't reconstruct terms if there was no change in sub terms - use the original term instead
- remove reconstruction of binders in the full reduction version (and call it "almost full reduction")

I also implemented your editorial suggestions except for the warning on the "almost full reduction" method since it should be save. We can have a full reduction version with 8.17.  Jason gave some hints how one could work around the issue by introducing an extra cast - essentially produce `forall (A : Type) (a : (A:Type)), a=a` instead of `forall (A:Type) (a:A), a=a`. But I would refrain from such hacks unless we see we need it.

andrew-appel commented:
I get the following strange behavior.  Typically in Coq I would write,

```
Require Import List.
Import ListNotations.
```

But mentioning compute_tactics_ltac2 interferes with that:

```
Require List.
Require compute_tactics_ltac2.
Locate List. (* prints:
  Module Ltac2.List
  Module Coq.Lists.List
  (shorter name to refer to it in current context is Lists.List)
*)
Import List. 
Import ListNotations.  (* fails: Cannot find module ListNotations. *)
```

I am not even Importing compute_tactics_ltac2, and still it drags in Ltac2.List in a way that shadows Coq.Lists.List.

MSoegtropIMC commented:
I guess the issue is this:
```
Require List.
Locate List.
(* => Module Coq.Lists.List *)

Require Ltac2.Ltac2.
Locate List.
(* =>
Module Ltac2.List
Module Coq.Lists.List
  (shorter name to refer to it in current context is Lists.List)
*)
```

So Requiring a module puts it into the list of things one can Import - and apparently the last one required is then at the top of the list. As the Locate command says, you have to use `Import Lists.List`).

That is indeed unfortunate. What I don't quite understand is why an unspecified `Require List` does not give an error that it is not unique - maybe core Coq stuff takes precedence. Thinking about it I darkly remember the reason that Ltac2 is under `user-contrib` is exactly this module name clash issue and that Coq core takes precedence - until one wilfully requires Ltac2.
